### PR TITLE
Revert "Revert agent/helm: Deprecate --kpr=partial|strict|disabled and use --kpr=true|false instead"

### PR DESCRIPTION
**Pull request by brb**
#26493 reverted #26036, because it broke ci-e2e in #26373 PR. However, the PR didn't rebase against the latest main branch which included #26036. The failure was the following:

> Error: Unable to install Cilium: execution error at (cilium/templates/cilium-configmap.yaml:58:5): kubeproxyReplacement must be explicitly set to a valid value (disabled, partial, or strict) to continue.

Because ci-e2e (and other GHA workflows due to security reasons) takes Helm charts from a PR, and the ci-e2e workflow file from the main, the new KPR value was not recognized by the PR's Helm charts validation. A simple rebase in the PR would have resolved the issue.

cc @joestringer

### CODE REVIEW

**Review by jrajahalme — APPROVED** [review]
Lets get this in again :-)

**Comment by jrajahalme** [comment]
@brb Consider adding this here:
```
diff --git a/install/kubernetes/cilium/templates/cilium-configmap.yaml b/install/kubernetes/cilium/templates/cilium-configmap.yaml
index b3a12a6c86..3fe2079359 100644
--- a/install/kubernetes/cilium/templates/cilium-configmap.yaml
+++ b/install/kubernetes/cilium/templates/cilium-configmap.yaml
@@ -50,6 +50,12 @@
   {{- $defaultKubeProxyReplacement = "disabled" -}}
 {{- end -}}
 
+{{- /* Default values when 1.14 was initially deployed */ -}}
+{{- if semverCompare ">=1.14" (default "1.14" .Values.upgradeCompatibility) -}}
+  {{- /* KPR default for 1.14 needed to override earlier version defaults set above when upgradeCompatibility is not specified */ -}}
+  {{- $defaultKubeProxyReplacement = "false" -}}
+{{- end -}}
+
 {{- $ipam := (coalesce .Values.ipam.mode $defaultIPAM) -}}
 {{- $bpfCtTcpMax := (coalesce .Values.bpf.ctTcpMax $defaultBpfCtTcpMax) -}}
 {{- $bpfCtAnyMax := (coalesce .Values.bpf.ctAnyMax $defaultBpfCtAnyMax) -}}
```
Without this new installs will have kpr=disabled if `upgradeCompatibility` is not explicitly given?

Alternatively, maybe we should bump all the default values for `upgradeCompatibility` to "1.14"?

For example:
```
{{- if semverCompare ">=1.12" (default "1.12" .Values.upgradeCompatibility) -}}
```
-> 
```
{{- if semverCompare ">=1.12" (default "1.14" .Values.upgradeCompatibility) -}}
```
That way the old defaults would only take effect when `upgradeCompatibility` is explicitly given?

**Review by joestringer — APPROVED** [review]

**Comment by brb** [comment]
> Consider adding this here:

Good suggestion. I will do that in a follow-up PR once this gets merged.

**Comment by asauber** [comment]
@jrajahalme Can you explain a bit more what you mean here?

>     {{- if semverCompare ">=1.12" (default "1.14" .Values.upgradeCompatibility) -}}
> 
> That way the old defaults would only take effect when upgradeCompatibility is explicitly given?

Are there any cases currently where changing the default would result in different values being used for a 1.14 deployment?

Is this mostly to prepare for the cases where there are more recent defaults than 1.12 defaults?

**Comment by jrajahalme** [comment]
> @jrajahalme Can you explain a bit more what you mean here?
> 
> > ```
> > {{- if semverCompare ">=1.12" (default "1.14" .Values.upgradeCompatibility) -}}
> > ```
> > That way the old defaults would only take effect when upgradeCompatibility is explicitly given?
> 
> Are there any cases currently where changing the default would result in different values being used for a 1.14 deployment?
> 
> Is this mostly to prepare for the cases where there are more recent defaults than 1.12 defaults?

I was simply mistaken here, as changing the default to a later version does not change the value of the if condition here, i.e., both `semverCompare ">=1.12" "1.14")` and `semverCompare ">=1.12" "1.12")` are `true`!

To get what I was after, we would need to specify these as:
```
{{- if semverCompare ">=1.12" (default "0.0" .Values.upgradeCompatibility) -}}
```
then none of the old defaults would take effect if `upgradeCompatibility` is not given.

Currently, when `upgradeCompatibility` is not given, each of these if conditions is `true`, and we set `kpr=probe`, for example. This means that we have to be careful to have later if statements for later versions that then override these. I had to add the one for 1.14 I mentioned above to make Envoy Config work in #25901.

kpr is the only one I was made aware of, I have not analysed any of the other defaults set here.

File: daemon/cmd/kube_proxy_replacement.go [hunk 1]
```diff
@@ -45,10 +45,19 @@ import (
...
+		option.Config.KubeProxyReplacement != option.KubeProxyReplacementDisabled &&
+		option.Config.KubeProxyReplacement != option.KubeProxyReplacementTrue &&
+		option.Config.KubeProxyReplacement != option.KubeProxyReplacementFalse {
 		return fmt.Errorf("Invalid value for --%s: %s", option.KubeProxyReplacement, option.Config.KubeProxyReplacement)
 	}
 
+	if option.Config.KubeProxyReplacement == option.KubeProxyReplacementStrict ||
+		option.Config.KubeProxyReplacement == option.KubeProxyReplacementPartial ||
+		option.Config.KubeProxyReplacement == option.KubeProxyReplacementDisabled {
+		log.Warnf("Deprecated value for --%s: %s (use either \"true\", or \"false\")", option.KubeProxyReplacement, option.Config.KubeProxyReplacement)
```
Comment (pchaigno) [inline daemon/cmd/kube_proxy_replacement.go:57]: @brb I think you forgot to update the ginkgo tests. This warning is now popping up in many of them.

Comment (brb) [reply daemon/cmd/kube_proxy_replacement.go:57]: Thanks for reporting! https://github.com/cilium/cilium/pull/29695.

**Pull request merged**